### PR TITLE
parser: replace Expression with ExpressionListOpt

### PR DESCRIPTION
**Pull request by zhexuany**
The goal of this PR is let illegal argument pass into executor stage. Let executor determine the number of argument is correct or not.

### CODE REVIEW

**Comment by shenli** [comment]
@zhexuany Please fix the CI.

**Comment by coocood** [comment]
@zhexuany 
Great Job!
If you view the diff, you can see there are some items not aligned.
In `parser.y` we use tab instead of white space to align.
We didn't find a tool to format .y file, so we have to verify it by human eyes for now.

**Comment by zhexuany** [comment]
@coocood There is wired bug. I modify `ABS(Expression)`  to `ABS(ExpressionListOpt)` passed test but raise painc. 
~~~
2017/04/24 02:17:25 session.go:569: [warning] [1] compile error:
[expression:1582]Incorrect parameter count in the call to native function 'abs'
select abs(1, 1)
2017/04/24 02:17:25 conn.go:367: [warning] [1] dispatch error:
id:1, addr:127.0.0.1:55439 status:2, collation:utf8_general_ci, user:root
select abs(1, 1)
/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/expression/builtin.go:112: [expression:1582]Incorrect parameter count in the call to native function 'abs'
/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/expression/builtin_math.go:97:
/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/expression/scalar_function.go:74:
/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/plan/expression_rewriter.go:72:
/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/plan/logical_plan_builder.go:303:
/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/plan/optimizer.go:69:
/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/executor/compiler.go:43:
/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/tidb.go:173:
/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/session.go:571:
/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/server/conn.go:674:
2017/04/24 02:17:27 conn.go:336: [error] lastCmd select abs(), runtime error: index out of range, goroutine 75 [running]:
github.com/pingcap/tidb/server.(*clientConn).Run.func1(0xc4207c2580)
	/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/server/conn.go:334 +0xcd
panic(0x1b1bd60, 0x252dae0)
	/usr/local/Cellar/go/1.8/libexec/src/runtime/panic.go:489 +0x2cf
github.com/pingcap/tidb/plan.(*typeInferrer).handleFuncCallExpr(0xc42043e120, 0xc4207b8460)
	/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/plan/typeinferer.go:302 +0x126f
github.com/pingcap/tidb/plan.(*typeInferrer).Leave(0xc42043e120, 0x2260340, 0xc4207b8460, 0x2260340, 0xc4207b8460, 0xc4203b8600)
	/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/plan/typeinferer.go:69 +0xf49
github.com/pingcap/tidb/ast.(*FuncCallExpr).Accept(0xc4207b8460, 0x2258480, 0xc42043e120, 0x2260b00, 0xc4204288a0, 0xc4203b7600)
	/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/ast/functions.go:305 +0x1a2
github.com/pingcap/tidb/ast.(*SelectField).Accept(0xc4204288a0, 0x2258480, 0xc42043e120, 0x22602c0, 0xc42043e090, 0x0)
	/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/ast/dml.go:294 +0xfb
github.com/pingcap/tidb/ast.(*FieldList).Accept(0xc42043e090, 0x2258480, 0xc42043e120, 0x2260b40, 0xc4207bfa70, 0xc4203b7700)
	/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/ast/dml.go:318 +0xc6
github.com/pingcap/tidb/ast.(*SelectStmt).Accept(0xc4207bfa70, 0x2258480, 0xc42043e120, 0x12638c3, 0xc420319fc0, 0x2260b40)
	/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/ast/dml.go:505 +0x5f9
github.com/pingcap/tidb/plan.InferType(0xc4203bbd80, 0x2260b40, 0xc4207bfa70, 0x100fe18, 0x20)
	/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/plan/typeinferer.go:35 +0x97
github.com/pingcap/tidb/plan.Optimize(0x226d8e0, 0xc4207c2630, 0x2260b40, 0xc4207bfa70, 0x2826ba0, 0xc4201b9050, 0x0, 0x0, 0x27c0960, 0x0)
	/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/plan/optimizer.go:57 +0x78
github.com/pingcap/tidb/executor.(*Compiler).Compile(0xc4203b7940, 0x226d8e0, 0xc4207c2630, 0x2826c60, 0xc4207bfa70, 0x1059c45, 0x1086ba2, 0xc420242360, 0xc4203bbd80)
	/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/executor/compiler.go:41 +0x1ae
github.com/pingcap/tidb.Compile(0x226d8e0, 0xc4207c2630, 0x2826c60, 0xc4207bfa70, 0x0, 0x0, 0x0, 0xc420168a10)
	/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/tidb.go:171 +0x57
github.com/pingcap/tidb.(*session).Execute(0xc4207c2630, 0xc420169ac1, 0xc, 0xc4203b7c08, 0x101094c, 0xc4203b7c08, 0x1010f22, 0xc42034c030)
	/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/session.go:567 +0x515
github.com/pingcap/tidb/server.(*TiDBContext).Execute(0xc420343ec0, 0xc420169ac1, 0xc, 0x0, 0x0, 0xc4203b7cd8, 0x106ef48, 0xc420765918)
	/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/server/driver_tidb.go:193 +0x62
github.com/pingcap/tidb/server.(*clientConn).handleQuery(0xc4207c2580, 0xc420169ac1, 0xc, 0x0, 0x0)
	/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/server/conn.go:671 +0x8d
github.com/pingcap/tidb/server.(*clientConn).dispatch(0xc4207c2580, 0xc420169ac1, 0xd, 0xc, 0x0, 0x0)
	/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/server/conn.go:457 +0x42e
github.com/pingcap/tidb/server.(*clientConn).Run(0xc4207c2580)
	/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/server/conn.go:352 +0x123
github.com/pingcap/tidb/server.(*Server).onConn(0xc4203464b0, 0x226cc40, 0xc420194448)
	/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/server/server.go:229 +0x148
created by github.com/pingcap/tidb/server.(*Server).Run
	/Users/zhexuany/repo/go/src/github.com/pingcap/tidb/server/server.go:178 +0xc5

2017/04/24 02:17:27 server.go:212: [info] [1] close connection
^C2017/04/24 02:21:26 main.go:162: [info] Got signal [2] to exit.
~~~

By change the order of execution in optimizer.go (see details https://github.com/pingcap/tidb/pull/3124/commits/8f8c0fb70dab7ed73b1c2ab1cdcd5f15103111ea#diff-cb003f7e1a477bea600e8e4f0de5858bL56), this bug is gone.  I am not very familiar with this area. Could you take a look?

**Comment by zhexuany** [comment]
@coocood  See my latest commit. Should I fix it here or fill another PR?

**Comment by coocood** [comment]
@zhexuany 
This is not the right way to fix it.
The panic is raised by `typeInferrer`.

**Comment by coocood** [comment]
@zhexuany 
The problem is that the typeinfer process run before argument check.
When argument count is specified by `parser`, it is ok, but when you change it to ExpressionList, wrong argument leads to panic.

**Comment by zhexuany** [comment]
@coocood  Yeah I agrees with you. That is I undo my changes on optimizer. 538081b fixed this panic by rewriting abs().

**Comment by coocood** [comment]
Please add a test case that can reproduce the panic.

**Comment by zhexuany** [comment]
@coocood its gone. Check here https://github.com/pingcap/tidb/pull/3124/commits/538081b9595584d04817009914f0853f6412decb#diff-dfd54d70dfa7a948dc6619c861f9f9b3L105

The question is: I do it here or file another PR?

File: parser/parser.y [hunk 1]
```diff
@@ -2782,171 +2782,124 @@ FunctionCallKeyword:
...
+|	"ASIN" '(' ExpressionListOpt ')'
 	{
-		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: []ast.ExprNode{$3.(ast.ExprNode)}}
+		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: $3.([]ast.ExprNode)}
 	}
-|	"ATAN" '(' Expression ',' Expression ')'
+|	"ATAN" '(' ExpressionListOpt ')'
 	{
-		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: []ast.ExprNode{$3.(ast.ExprNode)}}
+    $$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: $3.([]ast.ExprNode)}
```
Comment (tiancaiamao) [inline parser/parser.y]: format it.

File: parser/parser.y [hunk 2]
```diff
@@ -2964,104 +2917,82 @@ FunctionCallNonKeyword:
...
 	{
+
 		$$ = &ast.FuncCallExpr{
-			FnName: model.NewCIStr($1),
-			Args: []ast.ExprNode{
-				$3.(ast.ExprNode),
-				$5.(ast.ExprNode),
-				ast.NewValueExpr("DAY"),
-			},
+            FnName: model.NewCIStr($1),
```
Comment (tiancaiamao) [inline parser/parser.y]: format it.

File: parser/parser.y [hunk 3]
```diff
@@ -3577,33 +3406,29 @@ FunctionCallNonKeyword:
...
 		}
 	}
-|	"TRUNCATE" '(' Expression ',' Expression ')'
+|	"TRUNCATE" '(' ExpressionListOpt ')'
 	{
-		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: []ast.ExprNode{$3.(ast.ExprNode), $5.(ast.ExprNode)}}
-	}
-|	"UPPER" '(' Expression ')'
+		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: $3.([]ast.ExprNode)}
+ }
```
Comment (tiancaiamao) [inline parser/parser.y]: ditto

File: parser/parser.y [hunk 4]
```diff
@@ -3577,33 +3406,29 @@ FunctionCallNonKeyword:
...
 	}
-|	"UCASE" '(' Expression ')'
+|	"UCASE" '(' ExpressionListOpt ')'
 	{
-		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: []ast.ExprNode{$3.(ast.ExprNode)}}
+		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: $3.([]ast.ExprNode)}
 	}
 |	"UNIX_TIMESTAMP" '(' ')'
-	{
+  {
```
Comment (tiancaiamao) [inline parser/parser.y]: ditto

File: parser/parser.y [hunk 5]
```diff
@@ -3577,33 +3406,29 @@ FunctionCallNonKeyword:
...
 		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1)}
 	}
-|	"UNIX_TIMESTAMP" '(' Expression ')'
-	{
-		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: []ast.ExprNode{$3.(ast.ExprNode)}}
+|	"UNIX_TIMESTAMP" '(' ExpressionList ')'
+  {
```
Comment (tiancaiamao) [inline parser/parser.y]: ditto

**Comment by zhexuany** [comment]
@coocood @tiancaiamao PTAL

File: parser/parser.y (hunk 3 above)
Comment (zhexuany) [inline parser/parser.y]: DONE

**Comment by coocood** [comment]
LGTM

File: parser/parser.y [hunk 6]
```diff
@@ -3163,193 +3065,149 @@ FunctionCallNonKeyword:
...
+		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: $3.([]ast.ExprNode)}
+  }
+|	"MAKETIME" '(' ExpressionListOpt ')'
 	{
-		args := []ast.ExprNode{$3.(ast.ExprNode), $5.(ast.ExprNode), $7.(ast.ExprNode)}
-		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: args}
-	}
-|	"MAKE_SET" '(' ExpressionList ')'
+		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: $3.([]ast.ExprNode)}
+  }
```
Comment (tiancaiamao) [inline parser/parser.y]: ditto

File: parser/parser.y [hunk 7]
```diff
@@ -3163,193 +3065,149 @@ FunctionCallNonKeyword:
...
+		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: $3.([]ast.ExprNode)}
 	}
-|	"MONTH" '(' Expression ')'
+|	"MONTH" '(' ExpressionListOpt ')'
 	{
-		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: []ast.ExprNode{$3.(ast.ExprNode)}}
-	}
-|	"MONTHNAME" '(' Expression ')'
+		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: $3.([]ast.ExprNode)}
+  }
```
Comment (zhexuany) [inline parser/parser.y]: confused. It is already formatted.

File: parser/parser.y [hunk 8]
```diff
@@ -3163,193 +3065,149 @@ FunctionCallNonKeyword:
...
 	}
-|	"MAKEDATE" '(' Expression ',' Expression ')'
+|	"MAKEDATE" '(' ExpressionListOpt ')'
 	{
-		args := []ast.ExprNode{$3.(ast.ExprNode), $5.(ast.ExprNode)}
-		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: args}
-	}
-|	"MAKETIME" '(' Expression ',' Expression ',' Expression ')'
+		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: $3.([]ast.ExprNode)}
+  }
```
Comment (zhexuany) [inline parser/parser.y]: It is already formatted.

File: parser/parser.y (hunk 6 above)
Comment (zhexuany) [inline parser/parser.y]: It is already formatted.

**Comment by zhexuany** [comment]
@tiancaiamao PTAL.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**